overlord/state/copy.go

```diff
@@ -0,0 +1,122 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package state
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/snapcore/snapd/jsonutil"
+	"github.com/snapcore/snapd/osutil"
+)
+
+type checkpointOnlyBackend struct {
+	path string
+}
+
+func (b *checkpointOnlyBackend) Checkpoint(data []byte) error {
+	return osutil.AtomicWriteFile(b.path, data, 0600, 0)
+}
+
+func (b *checkpointOnlyBackend) EnsureBefore(d time.Duration) {
+	panic("cannot use EnsureBefore in checkpointOnlyBackend")
+}
+
+func (b *checkpointOnlyBackend) RequestRestart(t RestartType) {
+	panic("cannot use RequestRestart in checkpointOnlyBackend")
+}
+
+func copyData(subkeys []string, pos int, srcData map[string]*json.RawMessage, dstData map[string]interface{}) error {
+	raw, ok := srcData[subkeys[pos]]
```

Comment:
For sanity I think we could check that `pos >= 0 && pos < len(subkeys)` and return an error if it's not